src/NServiceBus.Core/Routing/TransportAddressTranslator.cs

```diff
@@ -0,0 +1,27 @@
+namespace NServiceBus.Features
+{
+    using System.Collections.Concurrent;
+    using Routing;
+    using Transport;
+
+    class TransportAddressTranslator
+    {
+        public TransportAddressTranslator(TransportInfrastructure transportInfrastructure)
+        {
+            this.transportInfrastructure = transportInfrastructure;
+        }
+
+        public string ToTransportAddress(EndpointInstance remoteEndpointInstance)
+        {
+            return cache.GetOrAdd(remoteEndpointInstance, Translate);
```

Comment:
If we're trying to save on allocations, this will allocate a delegate each call.

The following would prevent the delegate and still preserve the caching:

``` csharp
string value;
if (!cache.TryGetValue(remoteEndpointInstance, out value))
{
    value = cache.GetOrAdd(remoteEndpointInstance, Translate(remoteEndpointInstance));
}

return value;
```
At the expense of two TryGet calls when the cache is empty.